File: algospot/timetrip.cpp
#include <iostream>
#include <vector>
#include <functional>
#include <string>

using namespace std;
using pi = pair<int, int>;
int main() {
#ifdef _DEBUG
	freopen("timetrip.in", "r", stdin);
#endif
	int c, v, w, a, b, d;
	cin >> c;
	while (c--) {
		cin >> v >>  w;
		vector<vector<pi>> adj(v, vector<pi>());
		for (int l = 0; l < w; l++) {
			cin >> a >> b >> d;
			adj[a].emplace_back(b, d);
		}
		vector<vector<bool>> reachable(v, vector<bool>(v, false));
		for (int i = 0; i < v; i++) 
			reachable[i][i] = true;
		for (int a = 0; a < v; a++)
			for (auto& b : adj[a])
				reachable[a][b.first] = true;
		for (int a = 0; a < v; a++) 
			for (int i = 0; i < v; i++) 
				for (int j = 0; j < v; j++) 
					reachable[i][j] = reachable[i][j] or (reachable[i][a] and reachable[a][j]);
		int inf = 987654321;
		if (!reachable[0][1]) {
			cout << "UNREACHABLE" << endl;
			continue;
		}
		auto bellman2 = [&](int src, int target, function<int(int,int)> func) {
			vector<int> upper(v, inf);
			upper[src] = 0;
			for (int iter = 0; iter < v-1; iter++) 
				for (int here = 0; here < v; ++here)
					for (auto& k :adj[here]) {
						int there = k.first;
						int cost = k.second;
						upper[there] = min(upper[there], func(upper[here], cost));
					}
			for (int here = 0; here < v; ++here)
				for (auto& k : adj[here]) {
					int there = k.first;
					int cost = k.second;
					if (func(upper[here], cost) < upper[there])
						if (reachable[src][here] && reachable[here][target])
							return -inf;
				}
			return upper[target];
		};
		auto min =  bellman2(0, 1, plus<int>());
		auto max = -bellman2(0, 1, minus<int>());
		auto filter = [&](int i) { if (abs(i) > inf/2) return string("INFINITY"); else return to_string(i); };
		cout << filter(min) << " " << filter(max) << endl;
	}
	return 0;
}

File: algospot/timetrip.py
def bellman2(src, dest, op):
    inf = 1e10
    upper = [inf] * v
    upper[src] = 0
    for loop in range(v-1):
        for a in range(v):
            for b, d in adj[a]:
                upper[b] = min(upper[b], op(upper[a], d))
    for a in range(v):
        for b, d in adj[a]:
            if upper[b] > op(upper[a], d):
                if reachable[src][a] and reachable[b][dest]:
                    return -inf
    return upper[dest]


for case in range(int(input())):
    v, w = map(int, input().split())
    adj = {i: [] for i in range(v)}
    for _ in range(w):
        a, b, d = map(int, input().split())
        adj[a].append((b, d))
    reachable = [[False for _ in range(v)] for _ in range(v)]
    for i in range(v):
        reachable[i][i] = True
    for a in range(v):
        for b, d in adj[a]:
            reachable[a][b] = True
    for a in range(v):
        for i in range(v):
            for j in range(v):
                reachable[i][j] = reachable[i][j] or (reachable[i][a] and reachable[a][j])
    if not reachable[0][1]:
        print("UNREACHABLE")
    else:
        ans = [bellman2(0, 1, lambda a, b: a+b), -bellman2(0, 1, lambda a, b: a - b)]
        for n, i in enumerate(ans):
            if abs(i) > 1e5:
                ans[n] = "INFINITY"
        print(ans[0], ans[1])
